### ISSUE TITLE
Design an API for the MPA transition

### ISSUE THREAD

**Issue by vmpstr** [WICG/view-transitions#2]
The SPA transition case has a working API, but it's unclear what the MPA equivalent should look like.

We need to design that

**Comment by vmpstr** [comment]
I think we still need a prepare step, so that should remain the same. However, the start call should be replaced with something like

```
startOnNavigation(url, sharedElements: selectors);
```

I think we would want to pass the url here, which would initiate the navigation and sharedElement selectors will need to be a list of identifiers that can be used to identify elements on the destination page.

It's unclear whether these elements should be specially tagged (ie opt-in style, something like `sharedID = "foo"`) or just rely on regular css selectors.

One thing to note is that for SPA, we require paint containment right now, and I think we need that for the MPA case as well.

/cc @jakearchibald

**Comment by jakearchibald** [comment]
If possible, I think we should avoid triggering the navigation with a new API. There are lots of ways to navigate, some get special privileges, and some modify history state in a particular way. If we create a new API we'll likely lose some of those, or end up with an option bag to try and cover all of those cases.

https://github.com/WICG/app-history has a "navigate" event, which sounds like exactly what we'd need.

> It's unclear whether these elements should be specially tagged (ie opt-in style, something like `sharedID = "foo"`) or just rely on regular css selectors.

This makes me wonder if the MPA API would be better in CSS:

```css
@nav-transition {
  type: reveal;
  duration: 300ms;
}

.main-avatar {
  nav-transition-id: main-avatar;
}

.header {
  nav-transition-id: header;
}
```

Setting `nav-transition-id` could also enforce other CSS requirements, like paint containment.

I'm not sure how you'd specify different nav transitions for different destinations. Maybe you could do that with JS in the "navigate" event. Or, something like:

```css
@nav-transition {
  type: reveal;
  duration: 300ms;
  destination: url-prefix('/profile/');
}
```

…where the first `@nav-transition` block with the matching destination is used.

> One thing to note is that for SPA, we require paint containment right now, and I think we need that for the MPA case as well.

That seems ok, but what's the reason for the requirement?

**Comment by jakearchibald** [comment]
Some feedback I got: The 'destination' should be part of the @-rule. We also need to consider how the inclusion of a URL might make it harder to deliver these transitions in the form of a library.

**Comment by bramus** [comment]
@jakearchibald (thinking out loud here) How about some generic keywords such as `forwards` / `backwards`?

**Comment by nickcoury** [comment]
A few things that would be nice to have in this API:

1. Separate concepts of exit and enter animations.

One reason for using a SPA and animating transitions is to make apps feel immediately responsive.  If I'm understanding this proposal, it would wait until the next page is ready to execute `.start()/.startOnNavigation()` since it requires the next page's elements to animate.

What would be nice is a two-phase animation.  Allow `.exitAnimation()` to run immediately when the next page is requested, which could allow any of the shared elements to be animated immediately.  For example, slide in a header, slide a shared element from its current location to its expected location on the next page, or show a loading spinner/ghost card.  Ideally with enough control to specify their positions, transforms, easing curves, etc.

Then, a second `.enterAnimation()` will be called once the next page is rendered enough to load.  This could still look close to the proposed API, but using the shared elements as positioned in the exit animation.  This still allows animations to the final positions easily.

2. Additional control of the render timeline.

If elements are being animated between both pages, the ideal case would be to finish the transition as soon as all shared elements are available.  If the transition is too early, animations may look terrible if some shared elements aren't rendered yet.  If we wait until the entire page is rendered, it may add a lot of latency and make SET for MPA feel slow.

Perhaps something like [renderblocking](https://gist.github.com/xiaochengh/c29c6246ead0cbb72708f10baa89bde4) could be extended to work on elements in the body chunk and used here?

Another improvement to the timeline would be to add an additional hook `.beforeEnterAnimation()` that would run once the destination page is being received, but before it is rendered.  This would be a small window, but if an intermediate animation could happen on the GPU to animate the shared elements one additional time, it might significantly improve the user experience.  For example, the header sliding in mentioned in `.exitAnimation()` could happen during this phase instead, and the loading spinner removed.  That provides better visual feedback and perceived responsiveness for users.

**Comment by khushalsagar** [comment]
And to your point about back/forward navigations, this came up in issue #25 as well. And as you mentioned this gets complicated if the new page changes before the back navigation.

Our thinking here is that the developer should be able to configure the transition for a back navigation. [app-history](https://github.com/WICG/app-history/blob/main/README.md) would provide a good hook to know when a back navigation is initiated and configure based on current state of the new page.

**Comment by daKmoR** [comment]
anything we can do to help out here?

**Comment by Pyrolistical** [comment]
Isn't there a large overlap of MPA transitions and `<portal>`s? https://web.dev/hands-on-portals/

I know its not exactly the same thing as being able to carry the same element across MPA boundaries, but you can fake it with `<portals>`.

It's better to have two independent proposals that compose than to have redundant effort.

**Comment by jeremyroman** [comment]
I agree, and in fact many of us working on both proposals sit in the same room. I think an API for shared element transitions between pages makes sense, even though there may be some overlap with portals.

**Comment by yoavweiss** [comment]
> [!NOTE]
> 🤖 **This comment was written and posted by an AI agent** (Claude, running in the [Pi](https://github.com/badlogic/pi-mono) coding agent), acting on behalf of @yoavweiss while triaging this repository ahead of archiving it. The triage was derived from the current spec text and from open issues in `w3c/csswg-drafts`. **If it's wrong, please say so and it will be reopened** — no hard feelings, it's a bot.

**Resolved.** Cross-document (MPA) view transitions shipped and are specified in [css-view-transitions-2 § cross-document view transitions](https://drafts.csswg.org/css-view-transitions-2/#cross-document-view-transitions). The API that landed is declarative opt-in on both ends:

```css
@view-transition { navigation: auto; }
```

...with `pageswap` / `pagereveal` (in HTML) as the JS hooks for customising per-navigation. Closing as completed.

---

<sub>**Where this work lives now:** View Transitions is specified in [css-view-transitions-1](https://drafts.csswg.org/css-view-transitions-1/) and [css-view-transitions-2](https://drafts.csswg.org/css-view-transitions-2/), maintained in [w3c/csswg-drafts](https://github.com/w3c/csswg-drafts/issues?q=is%3Aissue+is%3Aopen+label%3Acss-view-transitions-1%2Ccss-view-transitions-2). Please file new issues there.</sub>
